lib/stinger_core/inc/stinger_batch_insert.h

```diff
@@ -0,0 +1,571 @@
+    friend void stinger_batch_insert_edge_pairs<adapter, iterator>(stinger_t * G, iterator begin, iterator end);
+
+    // what 'iterator' points to
+    typedef typename std::iterator_traits<iterator>::value_type update;
+
+    // Result codes
+    // Before returning to caller, we subtract 10 to make return codes match up with functions like stinger_incr_edge
+    enum result_codes {
+        PENDING = 0,
+        EDGE_ADDED = 11,    // 1
+        EDGE_UPDATED = 10,  // 0
+        EDGE_NOT_ADDED = 9  // -1
+    };
+
+    // Set of functions to sort/filter a collection of updates by source vertex
+    struct source_funcs
+    {
+        static int64_t
+        get(const update &x){
+            return adapter::get_source(x);
+        }
+        static void
+        set(update &x, int64_t neighbor){
+            adapter::set_source(x, neighbor);
+        }
+        static bool
+        equals(const update &a, const update &b){
+            return adapter::get_source(a) == adapter::get_source(b);
+        }
+        static bool
+        compare(const update &a, const update &b){
+            return adapter::get_source(a) < adapter::get_source(b);
+        }
+        static bool
+        sort(const update &a, const update &b){
+            return (adapter::get_type(a) != adapter::get_type(b)) ? adapter::get_type(a) < adapter::get_type(b) :
+                   (adapter::get_source(a) != adapter::get_source(b)) ? adapter::get_source(a) < adapter::get_source(b) :
+                   (adapter::get_dest(a) != adapter::get_dest(b)) ? adapter::get_dest(a) < adapter::get_dest(b) :
+                   (adapter::get_time(a) != adapter::get_time(b)) ? adapter::get_time(a) < adapter::get_time(b) :
+                   false;
+        }
+    };
+
+    // Set of functions to sort/filter a collection of updates by destination vertex
+    struct dest_funcs
+    {
+        static int64_t
+        get(const update &x){
+            return adapter::get_dest(x);
+        }
+        static void
+        set(update &x, int64_t neighbor){
+            adapter::set_dest(x, neighbor);
+        }
+        static bool
+        equals(const update &a, const update &b){
+            return adapter::get_dest(a) == adapter::get_dest(b);
+        }
+        static bool
+        compare(const update &a, const update &b){
+            return adapter::get_dest(a) < adapter::get_dest(b);
+        }
+        static bool
+        sort(const update &a, const update &b){
+            return (adapter::get_type(a) != adapter::get_type(b)) ? adapter::get_type(a) < adapter::get_type(b) :
+                   (adapter::get_dest(a) != adapter::get_dest(b)) ? adapter::get_dest(a) < adapter::get_dest(b) :
+                   (adapter::get_source(a) != adapter::get_source(b)) ? adapter::get_source(a) < adapter::get_source(b) :
+                   (adapter::get_time(a) != adapter::get_time(b)) ? adapter::get_time(a) < adapter::get_time(b) :
```

Comment:
No, this one orders by dest then source, the other one orders by source then dest.